vm-virtio/src/fs.rs

```diff
@@ -0,0 +1,409 @@
+        .map_err(DeviceError::EpollCtl)?;
+        epoll::ctl(
+            epoll_fd,
+            epoll::ControlOptions::EPOLL_CTL_ADD,
+            self.kill_evt.as_raw_fd(),
+            epoll::Event::new(epoll::Events::EPOLLIN, u64::from(KILL_EVENT)),
+        )
+        .map_err(DeviceError::EpollCtl)?;
+
+        const EPOLL_EVENTS_LEN: usize = 100;
+        let mut events = vec![epoll::Event::new(epoll::Events::empty(), 0); EPOLL_EVENTS_LEN];
+
+        'epoll: loop {
+            let num_events =
+                epoll::wait(epoll_fd, -1, &mut events[..]).map_err(DeviceError::EpollWait)?;
+
+            for event in events.iter().take(num_events) {
+                let ev_type = event.data as u16;
+
+                match ev_type {
+                    VU_CALL_EVENT => {
+                        if let Err(e) = self.vu_call_evt.read() {
+                            error!("Failed to get queue event: {:?}", e);
+                            break 'epoll;
+                        } else {
+                            self.interrupt_status
+                                .fetch_or(INTERRUPT_STATUS_USED_RING as usize, Ordering::SeqCst);
+                            self.interrupt_evt
+                                .write(1)
+                                .map_err(DeviceError::FailedSignalingUsedQueue)?;
+                        }
+                    }
+                    KILL_EVENT => {
+                        debug!("KILL_EVENT received, stopping epoll loop");
+                        break 'epoll;
+                    }
+                    _ => {
+                        error!("Unknown event for virtio-fs");
+                    }
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub struct Fs {
+    vu: Master,
+    queue_sizes: Vec<u16>,
+    avail_features: u64,
+    acked_features: u64,
+    config_space: Vec<u8>,
+    kill_evt: Option<EventFd>,
+}
+
+impl Fs {
+    /// Create a new virtio-fs device.
+    pub fn new(path: &str, tag: &str, req_num_queues: usize, queue_size: u16) -> Result<Fs> {
+        // Calculate the actual number of queues needed.
+        let num_queues = NUM_QUEUE_OFFSET + req_num_queues;
+        // Connect to the vhost-user socket.
+        let mut master = Master::new(path, num_queues as u64).map_err(Error::VhostUserConnect)?;
+        // Retrieve available features only when connecting the first time.
+        let avail_features = master.get_features().map_err(Error::VhostUserGetFeatures)?;
+        // Create virtio device config space.
+        // First by adding the tag.
+        let mut config_space = tag.to_string().into_bytes();
+        config_space.resize(CONFIG_SPACE_SIZE, 0);
+        // And then by copying the number of queues.
+        let mut num_queues_slice: [u8; 4] = unsafe { transmute((req_num_queues as u32).to_be()) };
+        num_queues_slice.reverse();
+        config_space[CONFIG_SPACE_TAG_SIZE..CONFIG_SPACE_SIZE].copy_from_slice(&num_queues_slice);
+
```

Comment:
Can replace all this by `let num_queues_slice = req_num_queues.to_le_bytes()`

Reply:
Yes :)
Thanks for finding this `u32` function, this is better than this ugly code...